test/extensions/filters/listener/common/fuzz/listener_filter_fuzzer.cc

```diff
@@ -79,21 +87,33 @@ FuzzedHeader::FuzzedHeader(const test::extensions::filters::listener::FilterFuzz
 
   for (int i = 0; i < nreads_; i++) {
     data_.insert(data_.end(), input.data(i).begin(), input.data(i).end());
-    indices_.push_back(data_.size());
+    indices_.push_back(data_.size() - 1);
   }
 }
 
-Api::SysCallSizeResult FuzzedHeader::next(void* buffer, size_t length) {
-  if (done()) {           // End of stream reached
-    nread_ = nreads_ - 1; // Decrement to avoid out-of-range for last recv() call
+void FuzzedHeader::next() {
+  if (!done()) {
+    nread_++;
   }
-  memcpy(buffer, data_.data(), std::min(indices_[nread_], length));
-  return Api::SysCallSizeResult{static_cast<ssize_t>(indices_[nread_++]), 0};
 }
 
-bool FuzzedHeader::done() { return nread_ >= nreads_; }
+Api::SysCallSizeResult FuzzedHeader::read(void* buffer, size_t length, bool peek) {
+  const size_t len = std::min(size(), length); // Number of bytes to write
+  memcpy(buffer, data_.data() + index_, len);
+
+  if (!peek) {
+    // If not peeking, written bytes will be marked as read
+    index_ += len;
+  }
+
+  return Api::SysCallSizeResult{static_cast<ssize_t>(len), 0};
+}
+
+size_t FuzzedHeader::size() { return indices_[nread_] - index_ + 1; }
+
+bool FuzzedHeader::done() { return nread_ >= nreads_ - 1; }
```

Comment:
gotcha, i think this makes sense to me now. `nreads_` is the total number of reads, while `nread_` measures an index, so `nreads_` needs to be shifted down by one when using it to compare with an index or act as an index.